traffic_ops/traffic_ops_golang/trafficstats/trafficstats.go

```diff
@@ -0,0 +1,691 @@
+					found = true
+					break
+				}
+
+				if jsonWithUnixTimestamps.Equal(m) {
+					found = true
+					c.Unix = true
+					break
+				}
+			}
+
+			if !found {
+				e = fmt.Errorf("Failed to negotiate content; cannot produce output satisfying %s", accept)
+				code = http.StatusNotAcceptable
+				return
+			}
+		}
+	}
+
+	if limit, ok := i.Params["limit"]; ok {
+		lim, err := strconv.ParseUint(limit, 10, 64)
+		if err != nil {
+			e = errors.New("Invalid limit!")
+			return
+		}
+		c.Limit = &lim
+	}
+
+	if offset, ok := i.Params["offset"]; ok {
+		off, err := strconv.ParseUint(offset, 10, 64)
+		if err != nil {
+			e = errors.New("Invalid offset!")
+			return
+		}
+		c.Offset = &off
+	}
+
+	if orderby, ok := i.Params["orderby"]; ok {
+		if c.OrderBy = orderableFromString(orderby); c.OrderBy == nil {
+			e = errors.New("Invalid orderby!")
+			return
+		}
+	}
+
+	if c.Start, e = parseTime(i.Params["startDate"]); e != nil {
+		log.Errorf("Parsing startDate: %v", e)
+		e = errors.New("Invalid startDate!")
+		return
+	}
+
+	if c.End, e = parseTime(i.Params["endDate"]); e != nil {
+		log.Errorf("Parsing endDate: %v", e)
+		e = errors.New("Invalid endDate!")
+		return
+	}
+
+	if interval, ok := i.Params["interval"]; !ok {
+		c.Interval = DEFAULT_INTERVAL
+	} else if c.Interval = tc.TrafficStatsDurationFromString(interval); c.Interval == tc.InvalidDuration {
+		log.Errorf("Error parsing 'interval' query parameter: %v", e)
+		e = errors.New("Invalid interval!")
+		return
+	}
+
+	if ex, ok := i.Params["exclude"]; ok {
+		switch ex {
+		case "series":
+			c.ExcludeSeries = true
+		case "summary":
+			c.ExcludeSummary = true
+		default:
+			e = errors.New("Invalid exclude! Must be 'series' or 'summary'")
+			return
+		}
+	}
+
+	c.MetricType = i.Params["metricType"]
+	if _, ok := metricTypes[c.MetricType]; !ok {
+		e = fmt.Errorf("Unknown metric type: %s", c.MetricType)
+		return
+	}
+
+	var ok bool
+	if c.DeliveryService, ok = i.Params["deliveryServiceName"]; !ok {
+		if c.DeliveryService, ok = i.Params["deliveryService"]; !ok {
+			e = errors.New("You must specify deliveryService or deliveryServiceName!")
+			return
+		}
+
+		if dsID, err := strconv.ParseUint(c.DeliveryService, 10, 64); err != nil {
+			// sql.ErrNoRows does not *necessarily* mean the DS doesn't exist - an XMLID can simply
+			// be numeric, and so it was wrong to treat it as an ID in the first place.
+			xmlid := c.DeliveryService
+			if c.DeliveryService, err = getXMLIDFromID(dsID, i.Tx.Tx); err != nil && err != sql.ErrNoRows {
+				log.Errorf("Converting DSID to XMLID: %v", err)
+				e = errors.New("Internal Server Error")
+				code = http.StatusInternalServerError
+				return
+			} else if err == sql.ErrNoRows {
+				c.DeliveryService = xmlid
+			}
+		}
+	}
+
+	code = http.StatusOK
+	return
+}
+
+func GetDSStats(w http.ResponseWriter, r *http.Request) {
+	// Perl didn't require "interval", but it would only return summary data if it was not given
```

Comment:
nit but this method is pretty large I wonder if it should be split up?

Reply:
Sure

Reply:
I've moved some of the logic out into utilities, how's it look now?

Reply:
much better!!!